table/pit-cs-tree.go

```diff
@@ -0,0 +1,382 @@
+	return record
+}
+
+// Originally implemented as FindOrInsertOutRecord()
+func (e *nameTreePitEntry) GetOutRecords() []*PitOutRecord {
+	records := make([]*PitOutRecord, 0)
+	for _, value := range e.outRecords {
+		records = append(records, value)
+	}
+	return records
+}
+
+func (p *pitCsTreeNode) findExactMatchEntry(name *ndn.Name) *pitCsTreeNode {
+	if name.Size() > p.depth {
+		for _, child := range p.children {
+			if name.At(child.depth - 1).Equals(child.component) {
+				return child.findExactMatchEntry(name)
+			}
+		}
+	} else if name.Size() == p.depth {
+		return p
+	}
+	return nil
+}
+
+func (p *pitCsTreeNode) findLongestPrefixEntry(name *ndn.Name) *pitCsTreeNode {
+	if name.Size() > p.depth {
+		for _, child := range p.children {
+			if name.At(child.depth - 1).Equals(child.component) {
+				return child.findLongestPrefixEntry(name)
+			}
+		}
+	}
+	return p
+}
+
+func (p *pitCsTreeNode) fillTreeToPrefix(name *ndn.Name) *pitCsTreeNode {
+	curNode := p.findLongestPrefixEntry(name)
+	for depth := curNode.depth + 1; depth <= name.Size(); depth++ {
+		newNode := new(pitCsTreeNode)
+		newNode.component = name.At(depth - 1).DeepCopy()
+		newNode.depth = depth
+		newNode.parent = curNode
+		curNode.children = append(curNode.children, newNode)
+		curNode = newNode
+	}
+	return curNode
+}
+
+func (p *pitCsTreeNode) pruneIfEmpty() {
+	for curNode := p; curNode.parent != nil && len(curNode.children) == 0 && len(curNode.pitEntries) == 0 && curNode.csEntry == nil; curNode = curNode.parent {
+		// Remove from parent's children
+		for i, child := range curNode.parent.children {
+			if child == p {
+				if i < len(curNode.parent.children)-1 {
+					copy(curNode.parent.children[i:], curNode.parent.children[i+1:])
+				}
+				curNode.parent.children = curNode.parent.children[:len(curNode.parent.children)-1]
+				break
+			}
+		}
+	}
+}
+
+func (p *PitCsTree) generateNewPitToken() uint32 {
+	for {
+		token := rand.Uint32()
+		if _, ok := p.pitTokenMap[token]; !ok {
+			return token
+		}
+	}
+}
+
+func (p *PitCsTree) hashCsName(name *ndn.Name) uint64 {
+	return xxhash.Sum64String(name.String())
+}
+
+// FindDataExactMatch finds an exact matching entry in the CS (if any). If MustBeFresh is set to true in the Interest, only non-stale CS entries will be returned.
+// Originally implemented as FindMatchingDataCS()
+func (p *PitCsTree) FindDataExactMatch(interest *ndn.Interest) CsEntry {
+	node := p.root.findExactMatchEntry(interest.Name())
+	if node != nil {
+		if !interest.CanBePrefix() {
+			if node.csEntry != nil {
+				p.csReplacement.BeforeUse(node.csEntry.index, node.csEntry.data)
+			}
+			return node.csEntry
+		}
+		return node.findMatchingDataCSPrefix(interest) // TODO: is this a bug?
+		// Shouldn't we call this if node == nil, not if node != nil?
+	}
+	return nil
+}
+
+// Originally implemented as FindMatchingDataCS()
+func (p *PitCsTree) FindDataPrefixMatch(interest *ndn.Interest) CsEntry {
```

Comment:
Why you have this function? Data matching is a mixture of prefix match and exact match. See comments above.